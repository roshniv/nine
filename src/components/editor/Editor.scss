.editor {
    position: relative;
    padding: 0 16px;
    background-color: #f5f5f5;
    width: 760px;
    border-radius: 0 0 8px 8px;
    min-height: 337px;
    &:before {
        position: absolute;
        content: '';
        width: 100%;
        height: 4px;
        background-color: #1976d2;
        top: -12px;
        left: 0;
    }

    &__container {
        padding: 16px;
        display: flex;
        flex-direction: column-reverse;
    }

    &__edit-panel {
        width: 100%;
        margin-top: 32px;
    }
    &__display-panel {
        width: 100%;
    }
    &__verticle-divider {
        display: none;
    }
    &__control-field {
        display: flex;
        flex-direction: column;
        margin-bottom: 12px;
        &:last-child {
            padding-top: 8px;
            margin-bottom: 0;
        }
    }
    &__control-label {
        margin-bottom: 15px;
    }
    &__text-input {
        display: block;
        padding-left: 16px;
        font-size: 14px;
        height: 32px;
        border-radius: 4px;
        border: 1px solid #d3d3d3;
        transition: box-shadow 0.3s;
        &:focus {
            outline: none;
            box-shadow: 0 0 5px 2px #d3d3d3;
        }
    }
    &__radio-group {
        display: flex;
    }

    &__radio-btn {
        position: relative;
        cursor: pointer;
        width: 50%;
        padding-left: 32px;
    }
    &__radio-input {
        display: none;
        &:checked {
            ~ .editor__radio-icon {
                background-color: #1976d2;
                &:after {
                    display: block;
                }
            }
        }
    }
    &__radio-icon {
        position: absolute;
        top: 0;
        left: 0;
        bottom: 0;
        margin: auto;
        width: 16px;
        height: 16px;
        border: 1px solid #d3d3d3;
        border-radius: 50%;
        background-color: white;
        &:after {
            position: absolute;
			top: 0px;
			right: 0;
			bottom: 0;
			left: 0px;
			display: none;
			width: 6px;
			height: 6px;
			margin: auto;
			background: white;
			border-radius: 50%;
			content: '';
        }
    }

    // desktop
    @media (min-width: 768px) {
        &__container {
            padding: 62px;
            flex-direction: row;
        }
        &__edit-panel {
            width: 50%;
            padding-right: 32px;
            margin-top: 0;
        }
        &__display-panel {
            width: 50%;
            padding-left: 32px;
        }
        &__verticle-divider {
            display: block;
            margin: 16px 0;
            width: 1px;
            background-color: #d3d3d3;
        }
    }
}
