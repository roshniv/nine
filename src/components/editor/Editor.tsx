import React, { useState } from 'react';
import Widget from '../widget/Widget';
import './Editor.scss';

const Editor: React.FC = () => {
  // Assumption:
  // Title is not set as default
  // Temperature default unit is Celsius
  // Wind is shown in default
  const [isCelsius, setIsCelsius] = useState(true as boolean);
  const [showWind, setShowWind] = useState(true as boolean);

  return (
    <div className="editor">
      <div className="editor__container">
        <div className="editor__edit-panel">
          <div className="editor__control-field">
            <span className="editor__control-label">Title</span>
            <input
              type="text"
              className="editor__text-input"
              placeholder="Title of widget"
            />
          </div>
          <div className="editor__control-field">
            <span className="editor__control-label">Temperature</span>
            <div className="editor__radio-group">
              <label className="editor__radio-btn">
                <input
                  type="radio"
                  className="editor__radio-input"
                  name="temperature"
                  defaultChecked={isCelsius}
                  onChange={() => setIsCelsius(true)}
                />
                <span className="editor__radio-icon" />
                &#176;C
              </label>
              <label className="editor__radio-btn">
                <input
                  type="radio"
                  className="editor__radio-input"
                  name="temperature"
                  onChange={() => setIsCelsius(false)}
                />
                <span className="editor__radio-icon" />
                &#176;F
              </label>
            </div>
          </div>
          <div className="editor__control-field">
            <span className="editor__control-label">Wind</span>
            <div className="editor__radio-group">
              <label className="editor__radio-btn">
                <input
                  type="radio"
                  className="editor__radio-input"
                  name="wind"
                  defaultChecked={showWind}
                  onChange={() => setShowWind(true)}
                />
                <span className="editor__radio-icon" />
                On
              </label>
              <label className="editor__radio-btn">
                <input
                  type="radio"
                  className="editor__radio-input"
                  name="wind"
                  onChange={() => setShowWind(false)}
                />
                <span className="editor__radio-icon" />
                Off
              </label>
            </div>
          </div>
        </div>

        <span className="editor__verticle-divider" />

        <div className="editor__display-panel">
          <Widget
            isCelsius={isCelsius}
            showWind={showWind}
          />
        </div>
      </div>
    </div>
  );
};

export default Editor;
