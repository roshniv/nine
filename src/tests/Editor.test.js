/* eslint-disable max-len */
import React from 'react';
import {
  cleanup, render, screen, fireEvent,
} from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import Editor from '../components/editor/Editor';
import {
  mockGetlocationSuccess, mockAxiosGetSuccess, sydneyMockWeatherData,
} from './utils/mocks';

afterEach(cleanup);

jest.mock('axios');

// get html fragment for snapshot testing
let asFragment;

beforeEach(async () => {
  mockGetlocationSuccess();
  mockAxiosGetSuccess(sydneyMockWeatherData);

  await act(async () => {
    const renderedEditor = render(<Editor />);
    asFragment = renderedEditor.asFragment;
  });
});

describe('Editor component', () => {
  it('shows uppercase title in widget when typing in text input', async () => {
    const testTitle = 'title';
    fireEvent.change(screen.getByRole('textbox'), {
      target: { value: testTitle },
    });
    expect(asFragment()).toMatchSnapshot();
    expect(screen.getByTestId('widget-title')).toHaveTextContent(testTitle.toUpperCase());
  });

  it('has celsius and show wind checked in default', () => {
    expect(screen.getByLabelText('On')).toBeChecked();
    expect(screen.getByLabelText(/C$/)).toBeChecked();
  });

  it('convert kelvin to whole number celsius when check celsius', () => {
    const celsiusRadio = screen.getByLabelText(/C$/);
    const fahrenheitRadio = screen.getByLabelText(/F$/);
    fireEvent.click(fahrenheitRadio);
    fireEvent.click(celsiusRadio);

    expect(asFragment()).toMatchSnapshot();
    expect(celsiusRadio).toBeChecked();
    expect(screen.getByTestId('widget-info').querySelector('.widget__temperature')).toHaveTextContent(Math.round(sydneyMockWeatherData.main.temp - 273.15));
  });

  it('convert kelvin to whole number fahrenheit when check fahrenheit', () => {
    const fahrenheitRadio = screen.getByLabelText(/F$/);
    fireEvent.click(fahrenheitRadio);

    expect(asFragment()).toMatchSnapshot();
    expect(fahrenheitRadio).toBeChecked();
    expect(screen.getByTestId('widget-info').querySelector('.widget__temperature')).toHaveTextContent(Math.round((sydneyMockWeatherData.main.temp - 273.15) * 9 / 5 + 32));
  });

  it('should not show wind when wind off is checked', () => {
    const windOffRadio = screen.getByLabelText(/Off/);

    fireEvent.click(windOffRadio);

    expect(asFragment()).toMatchSnapshot();
    expect(windOffRadio).toBeChecked();
    expect(screen.getByTestId('widget-wind').querySelector('.widget__wind-status')).not.toBeInTheDocument();
  });

  it('should show wind when wind off is checked', () => {
    const windOnRadio = screen.getByLabelText(/On/);
    const windOffRadio = screen.getByLabelText(/Off/);
    fireEvent.click(windOffRadio);
    fireEvent.click(windOnRadio);

    expect(asFragment()).toMatchSnapshot();
    expect(windOnRadio).toBeChecked();
    expect(screen.getByTestId('widget-wind').querySelector('.widget__wind-status')).toBeInTheDocument();
  });
});
