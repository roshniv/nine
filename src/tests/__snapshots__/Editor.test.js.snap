// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor component convert kelvin to whole number celsius when check celsius 1`] = `
<DocumentFragment>
  <div
    class="editor"
  >
    <div
      class="editor__container"
    >
      <div
        class="editor__edit-panel"
      >
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Title
          </span>
          <input
            class="editor__text-input"
            placeholder="Title of widget"
            type="text"
          />
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Temperature
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                checked=""
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °C
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °F
            </label>
          </div>
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Wind
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                checked=""
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              On
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              Off
            </label>
          </div>
        </div>
      </div>
      <span
        class="editor__verticle-divider"
      />
      <div
        class="editor__display-panel"
      >
        <div
          class="widget"
        >
          <div
            class="widget__title"
            data-testid="widget-title"
          >
            TITLE
          </div>
          <div
            class="widget__content"
          >
            <img
              alt="weather"
              aria-label="weather-icon"
              class="widget__weather-img"
              src="day.svg"
            />
            <div
              class="widget__weather-info"
              data-testid="widget-info"
            >
              <div
                class="widget__city"
              >
                Sydney
              </div>
              <div
                class="widget__temperature"
              >
                12°
              </div>
              <div
                class="widget__wind"
                data-testid="widget-wind"
              >
                <span
                  class="widget__wind-status"
                >
                  Wind
                </span>
                <span
                  class="widget__wind-info"
                >
                  NW 1km/h
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Editor component convert kelvin to whole number fahrenheit when check fahrenheit 1`] = `
<DocumentFragment>
  <div
    class="editor"
  >
    <div
      class="editor__container"
    >
      <div
        class="editor__edit-panel"
      >
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Title
          </span>
          <input
            class="editor__text-input"
            placeholder="Title of widget"
            type="text"
          />
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Temperature
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °C
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °F
            </label>
          </div>
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Wind
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                checked=""
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              On
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              Off
            </label>
          </div>
        </div>
      </div>
      <span
        class="editor__verticle-divider"
      />
      <div
        class="editor__display-panel"
      >
        <div
          class="widget"
        >
          <div
            class="widget__title"
            data-testid="widget-title"
          >
            TITLE
          </div>
          <div
            class="widget__content"
          >
            <img
              alt="weather"
              aria-label="weather-icon"
              class="widget__weather-img"
              src="day.svg"
            />
            <div
              class="widget__weather-info"
              data-testid="widget-info"
            >
              <div
                class="widget__city"
              >
                Sydney
              </div>
              <div
                class="widget__temperature"
              >
                53°
              </div>
              <div
                class="widget__wind"
                data-testid="widget-wind"
              >
                <span
                  class="widget__wind-status"
                >
                  Wind
                </span>
                <span
                  class="widget__wind-info"
                >
                  NW 1km/h
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Editor component should not show wind when wind off is checked 1`] = `
<DocumentFragment>
  <div
    class="editor"
  >
    <div
      class="editor__container"
    >
      <div
        class="editor__edit-panel"
      >
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Title
          </span>
          <input
            class="editor__text-input"
            placeholder="Title of widget"
            type="text"
          />
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Temperature
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                checked=""
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °C
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °F
            </label>
          </div>
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Wind
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              On
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              Off
            </label>
          </div>
        </div>
      </div>
      <span
        class="editor__verticle-divider"
      />
      <div
        class="editor__display-panel"
      >
        <div
          class="widget"
        >
          <div
            class="widget__title"
            data-testid="widget-title"
          >
            TITLE
          </div>
          <div
            class="widget__content"
          >
            <img
              alt="weather"
              aria-label="weather-icon"
              class="widget__weather-img"
              src="day.svg"
            />
            <div
              class="widget__weather-info"
              data-testid="widget-info"
            >
              <div
                class="widget__city"
              >
                Sydney
              </div>
              <div
                class="widget__temperature"
              >
                12°
              </div>
              <div
                class="widget__wind"
                data-testid="widget-wind"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Editor component should show wind when wind off is checked 1`] = `
<DocumentFragment>
  <div
    class="editor"
  >
    <div
      class="editor__container"
    >
      <div
        class="editor__edit-panel"
      >
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Title
          </span>
          <input
            class="editor__text-input"
            placeholder="Title of widget"
            type="text"
          />
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Temperature
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                checked=""
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °C
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °F
            </label>
          </div>
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Wind
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                checked=""
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              On
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              Off
            </label>
          </div>
        </div>
      </div>
      <span
        class="editor__verticle-divider"
      />
      <div
        class="editor__display-panel"
      >
        <div
          class="widget"
        >
          <div
            class="widget__title"
            data-testid="widget-title"
          >
            TITLE
          </div>
          <div
            class="widget__content"
          >
            <img
              alt="weather"
              aria-label="weather-icon"
              class="widget__weather-img"
              src="day.svg"
            />
            <div
              class="widget__weather-info"
              data-testid="widget-info"
            >
              <div
                class="widget__city"
              >
                Sydney
              </div>
              <div
                class="widget__temperature"
              >
                12°
              </div>
              <div
                class="widget__wind"
                data-testid="widget-wind"
              >
                <span
                  class="widget__wind-status"
                >
                  Wind
                </span>
                <span
                  class="widget__wind-info"
                >
                  NW 1km/h
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Editor component shows uppercase title in widget when typing in text input 1`] = `
<DocumentFragment>
  <div
    class="editor"
  >
    <div
      class="editor__container"
    >
      <div
        class="editor__edit-panel"
      >
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Title
          </span>
          <input
            class="editor__text-input"
            placeholder="Title of widget"
            type="text"
          />
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Temperature
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                checked=""
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °C
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="temperature"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              °F
            </label>
          </div>
        </div>
        <div
          class="editor__control-field"
        >
          <span
            class="editor__control-label"
          >
            Wind
          </span>
          <div
            class="editor__radio-group"
          >
            <label
              class="editor__radio-btn"
            >
              <input
                checked=""
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              On
            </label>
            <label
              class="editor__radio-btn"
            >
              <input
                class="editor__radio-input"
                name="wind"
                type="radio"
              />
              <span
                class="editor__radio-icon"
              />
              Off
            </label>
          </div>
        </div>
      </div>
      <span
        class="editor__verticle-divider"
      />
      <div
        class="editor__display-panel"
      >
        <div
          class="widget"
        >
          <div
            class="widget__title"
            data-testid="widget-title"
          >
            TITLE
          </div>
          <div
            class="widget__content"
          >
            <img
              alt="weather"
              aria-label="weather-icon"
              class="widget__weather-img"
              src="day.svg"
            />
            <div
              class="widget__weather-info"
              data-testid="widget-info"
            >
              <div
                class="widget__city"
              >
                Sydney
              </div>
              <div
                class="widget__temperature"
              >
                12°
              </div>
              <div
                class="widget__wind"
                data-testid="widget-wind"
              >
                <span
                  class="widget__wind-status"
                >
                  Wind
                </span>
                <span
                  class="widget__wind-info"
                >
                  NW 1km/h
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
